File: 프로그래머스/연습문제/Lv2/숫자 블록/숫자 블록.cpp
#include <string>
#include <vector>
using namespace std;

#define MAX_DIV 10000000

vector<int> solution(long long begin, long long end) {
    vector<int> answer(end-begin+1, 1);
    if(begin == 1) answer[0] = 0;
    
    for(long long i=begin; i<=end; i++){
        // 1이면 0
        if(i == 1) continue;
        // 아니면 제일 큰 약수 찾기
        else{
            long long max_divisor = 1;
            for(long long j=2; j*j<=i; j++){
                if(!(i%j)){
                    if(i/j <= MAX_DIV){
                        answer[i-begin] = i/j;
                        break;
                    }
                    else max_divisor = j;
                }
                if(!(i%j) && i/j<=MAX_DIV){
                    answer[i-begin] = i/j;
                    break;
                }
            }
            
            if(answer[i-begin]==1 && max_divisor!=1)
                answer[i-begin] = max_divisor;
        }
    }
    
    return answer;
}

File: 프로그래머스/연습문제/Lv2/숫자 블록/숫자 블록.py
def solution(begin, end):
    MAX_DIV = 10000000

    answer = [1 for i in range(end-begin+1)]
    if begin == 1: answer[0] = 0

    for i in range(begin, end+1):
        if i == 1: continue
        else:
            max_divisor,j = 1,2
            while j**2 <= i:
                if i%j == 0:
                    if i//j <= MAX_DIV:
                        answer[i-begin] = i//j
                        break
                    else: max_divisor = j
                    
                j += 1

            if answer[i-begin]==1 and max_divisor!=1:
                    answer[i-begin] = max_divisor

    return answer
